File: ios/SsdidWallet/Domain/OpenId4Vci/CredentialOffer.swift
import Foundation

/// Transaction code requirement for pre-authorized code flow.
struct TxCodeRequirement {
    let inputMode: String
    let length: Int
    let description: String?
}

/// Parsed OpenID4VCI credential offer.
struct CredentialOffer {
    let credentialIssuer: String
    let credentialConfigurationIds: [String]
    let preAuthorizedCode: String?
    let txCode: TxCodeRequirement?
    let authorizationCodeGrant: Bool
    let issuerState: String?

    /// Parses a credential offer from a JSON string.
    static func parse(_ jsonString: String) throws -> CredentialOffer {
        guard let data = jsonString.data(using: .utf8),
              let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw OpenId4VciError.invalidOffer("Invalid JSON")
        }

        guard let issuer = obj["credential_issuer"] as? String else {
            throw OpenId4VciError.invalidOffer("Missing credential_issuer")
        }
        guard issuer.hasPrefix("https://") else {
            throw OpenId4VciError.invalidOffer("credential_issuer must be HTTPS: \(issuer)")
        }

        guard let configIds = obj["credential_configuration_ids"] as? [String],
              !configIds.isEmpty else {
            throw OpenId4VciError.invalidOffer("Missing or empty credential_configuration_ids")
        }

        guard let grants = obj["grants"] as? [String: Any] else {
            throw OpenId4VciError.invalidOffer("Missing grants")
        }

        let preAuthGrant = grants["urn:ietf:params:oauth:grant-type:pre-authorized_code"] as? [String: Any]
        let authCodeGrant = grants["authorization_code"] as? [String: Any]

        guard preAuthGrant != nil || authCodeGrant != nil else {
            throw OpenId4VciError.invalidOffer("Must have at least one grant type")
        }

        let preAuthCode = preAuthGrant?["pre-authorized_code"] as? String

        var txCode: TxCodeRequirement? = nil
        if let txCodeObj = preAuthGrant?["tx_code"] as? [String: Any] {
            txCode = TxCodeRequirement(
                inputMode: txCodeObj["input_mode"] as? String ?? "numeric",
                length: txCodeObj["length"] as? Int ?? 6,
                description: txCodeObj["description"] as? String
            )
        }

        let issuerState = authCodeGrant?["issuer_state"] as? String

        return CredentialOffer(
            credentialIssuer: issuer,
            credentialConfigurationIds: configIds,
            preAuthorizedCode: preAuthCode,
            txCode: txCode,
            authorizationCodeGrant: authCodeGrant != nil,
            issuerState: issuerState
        )
    }

    /// Parses a credential offer from a URI containing credential_offer or credential_offer_uri query parameter.
    static func parseFromUri(_ uriString: String) throws -> CredentialOffer {
        guard let components = URLComponents(string: uriString) else {
            throw OpenId4VciError.invalidOffer("Invalid URI: \(uriString)")
        }

        let queryItems = components.queryItems ?? []
        let offerJson = queryItems.first(where: { $0.name == "credential_offer" })?.value
        let offerUri = queryItems.first(where: { $0.name == "credential_offer_uri" })?.value

        if let offerJson = offerJson {
            return try parse(offerJson)
        } else if let offerUri = offerUri {
            guard offerUri.hasPrefix("https://") else {
                throw OpenId4VciError.invalidOffer("credential_offer_uri must be HTTPS")
            }
            throw OpenId4VciError.unsupported("By-reference offers require network fetch")
        } else {
            throw OpenId4VciError.invalidOffer("Missing credential_offer or credential_offer_uri")
        }
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/oid4vci/CredentialOffer.kt
package my.ssdid.wallet.domain.oid4vci

import kotlinx.serialization.json.*
import my.ssdid.sdk.domain.util.parseQueryParam

data class TxCodeRequirement(
    val inputMode: String,
    val length: Int,
    val description: String?
)

data class CredentialOffer(
    val credentialIssuer: String,
    val credentialConfigurationIds: List<String>,
    val preAuthorizedCode: String? = null,
    val txCode: TxCodeRequirement? = null,
    val authorizationCodeGrant: Boolean = false,
    val issuerState: String? = null
) {
    companion object {
        private val json = Json { ignoreUnknownKeys = true }

        fun parse(jsonString: String): Result<CredentialOffer> = runCatching {
            val obj = json.parseToJsonElement(jsonString).jsonObject

            val issuer = obj["credential_issuer"]?.jsonPrimitive?.content
                ?: throw IllegalArgumentException("Missing credential_issuer")
            require(issuer.startsWith("https://")) { "credential_issuer must be HTTPS: $issuer" }

            val configIds = obj["credential_configuration_ids"]?.jsonArray
                ?.map { it.jsonPrimitive.content }
                ?: throw IllegalArgumentException("Missing credential_configuration_ids")
            require(configIds.isNotEmpty()) { "credential_configuration_ids must not be empty" }

            val grants = obj["grants"]?.jsonObject
                ?: throw IllegalArgumentException("Missing grants")

            val preAuthGrant = grants["urn:ietf:params:oauth:grant-type:pre-authorized_code"]?.jsonObject
            val authCodeGrant = grants["authorization_code"]?.jsonObject

            require(preAuthGrant != null || authCodeGrant != null) { "Must have at least one grant type" }

            val preAuthCode = preAuthGrant?.get("pre-authorized_code")?.jsonPrimitive?.contentOrNull
            val txCodeObj = preAuthGrant?.get("tx_code")?.jsonObject
            val txCode = txCodeObj?.let {
                TxCodeRequirement(
                    inputMode = it["input_mode"]?.jsonPrimitive?.content ?: "numeric",
                    length = it["length"]?.jsonPrimitive?.int ?: 6,
                    description = it["description"]?.jsonPrimitive?.contentOrNull
                )
            }

            val issuerState = authCodeGrant?.get("issuer_state")?.jsonPrimitive?.contentOrNull

            CredentialOffer(
                credentialIssuer = issuer,
                credentialConfigurationIds = configIds,
                preAuthorizedCode = preAuthCode,
                txCode = txCode,
                authorizationCodeGrant = authCodeGrant != null,
                issuerState = issuerState
            )
        }

        fun parseFromUri(uriString: String): Result<CredentialOffer> = runCatching {
            val offerJson = parseQueryParam(uriString, "credential_offer")
            val offerUri = parseQueryParam(uriString, "credential_offer_uri")

            when {
                offerJson != null -> parse(offerJson).getOrThrow()
                offerUri != null -> {
                    require(offerUri.startsWith("https://")) { "credential_offer_uri must be HTTPS" }
                    throw UnsupportedOperationException("By-reference offers require network fetch")
                }
                else -> throw IllegalArgumentException("Missing credential_offer or credential_offer_uri")
            }
        }
    }
}
